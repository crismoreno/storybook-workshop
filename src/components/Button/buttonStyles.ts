import { css } from 'styled-components';
import { colorsLib } from '../../styles';

export const PrimaryButtonStyles = css`
  background-color: ${colorsLib.dark};
  color: ${colorsLib.clear};
`;

export const SecondaryButtonStyles = css`
  background-color: ${colorsLib.clear};
  color: ${colorsLib.dark};
`;
