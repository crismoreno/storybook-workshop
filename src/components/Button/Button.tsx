import React from 'react';
import { ReactNode } from 'react';
import styled from 'styled-components';
import '../../styles';
import { PrimaryButtonStyles, SecondaryButtonStyles } from './buttonStyles';

type ButtonTypes = 'primary' | 'secondary';

export type ButtonProps = {
  children: ReactNode;
  rounded?: boolean;
  buttonType?: ButtonTypes;
  disabled?: boolean;
  onClick?: () => void;
  icon?: ReactNode;
};

type StyledButtonProps = {
  $rounded?: boolean;
  $buttonType?: ButtonTypes;
  $disabled?: boolean;
};

const renderChildrenWithIcon = (
  children: ReactNode,
  icon: ReactNode
): JSX.Element => (
  <span className="inline-flex items-center">
    {children}
    <span className="ml-2 w-5">{icon}</span>
  </span>
);

const getStyleBasedOnType = (customType: ButtonTypes | undefined) => {
  if (customType === 'secondary') return SecondaryButtonStyles;
  return PrimaryButtonStyles;
};

const Button = ({
  children,
  icon,
  rounded,
  buttonType,
  disabled,
  onClick,
}: ButtonProps): JSX.Element => (
  <StyledButton
    onClick={onClick}
    $rounded={rounded}
    $disabled={disabled}
    $buttonType={buttonType}
  >
    {icon ? renderChildrenWithIcon(children, icon) : children}
  </StyledButton>
);

const StyledButton = styled.a<StyledButtonProps>`
  font-family: 'Source Sans Pro', sans-serif;
  cursor: pointer;
  border-radius: 4px;
  border: none;
  padding: 5px 10px;

  ${({ $buttonType }) => getStyleBasedOnType($buttonType)}
  ${({ $rounded }) =>
    $rounded &&
    `
			border-radius: 30px;
		`}
`;

export default Button;
