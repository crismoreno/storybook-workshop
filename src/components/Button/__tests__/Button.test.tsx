import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import type { RenderResult } from '@testing-library/react';

import Button from '../Button';

describe('Button', () => {
  const defaultProps = {};

  const getComponent = (props = defaultProps) => (
    <Button {...props}>Test Button</Button>
  );

  const setup = (props = defaultProps): RenderResult =>
    render(getComponent(props));

  it('Should render correctly', () => {
    const div = document.createElement('div');
    ReactDOM.render(getComponent(), div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot('Default');
  });
});
