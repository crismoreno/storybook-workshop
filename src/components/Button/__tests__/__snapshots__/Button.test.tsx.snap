// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should match snapshot: Default 1`] = `
<div>
  <a
    class="sc-bdnyFh cybjqN"
  >
    Test Button
  </a>
</div>
`;
